Speculative PR: Make default ExceptionHandler more informative #887

tkroman opened the pull request:
Comments on @jrudolph's suggestion to log stack traces completely ([as stated in the original issue](https://github.com/akka/akka-http/issues/887#issuecomment-281658746)) are welcome.

typesafehub-validator commented:
Hi @tkroman,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

jrudolph commented:
OK TO TEST

jlprat commented on akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala:
```diff
@@ -43,7 +43,15 @@ object ExceptionHandler {
+             |Completing with {} response.
+             |Stack traces aren't printed by default.
+             |To customize this behavior, consider implementing a custom ExceptionHandler
+             |(http://doc.akka.io/docs/akka-http/current/scala/http/routing-dsl/exception-handling.html)"""
```
I would not add the link to the docs in here. Structure may change, or hosted somewhere else, and this is the last thing you remember to modify when this happens.
Furthermore, the Java `ExceptionHandler` is just a wrapper on top of the Scala one, so even for Java API users it will point to the Scala documentation page.

tkroman replied:
If we decide that stack trace logging should be added, the last 3 lines become irrelevant and will be removed. Otherwise, I'll just remove the doc link, maybe restating the suggestion about providing a custom handler.

tkroman commented:
@jrudolph I changed the message and added error logging, which turned out to cause more changes that I originally expected (mostly in tests though). Do you think this is what we need? If my opinion counts, I like this better than a terse one-liner.

The latest build failed with 
`*** Error in `/usr/lib/jvm/java-1.8.0-openjdk/bin/java': free(): invalid pointer: 0x00000007c137b4f8 ***`, that's weird.

jlprat commented:
PLS BUILD

The error was probably some termporary hiccup on the Jenkins VM

jlprat commented:
Something that came to my mind, would it make sense to explain in the docs that the default behavior for the `ExceptionHandler` is to print the stack trace, plus an informative message?

jlprat commented on akka-http/src/main/scala/akka/http/scaladsl/server/ExceptionHandler.scala:
```diff
@@ -23,6 +23,10 @@ trait ExceptionHandler extends ExceptionHandler.PF {
 
 object ExceptionHandler {
   type PF = PartialFunction[Throwable, Route]
+  private[http] lazy val errorMessageTemplate: String = {
```
Why is this lazy? Not that is wrong, just wondering.

tkroman replied:
Actually, thanks for noticing this - I don't see a reason for that. I guess that `lazy` just slipped there. Also, in your comment above, WDYM by "plus an informative message"? Is it "add an informative message to logs" or "mention ExceptionHandler's informative message in docs"? If it's latter - yes, sure; if the former - do you have any suggestions about what to add to the current message?

jlprat replied:
Sorry, I meant "mention ExceptionHandler's informative message in docs" :smiley: So basically document your changes in the docs.

tkroman commented:
@jlprat I added some docs, but the wording in a couple of places seems a little off to me. Also, is it OK to include links to external resources? I added a link to the well-known 'The Exceptional Performance of Lil' Exception' post to back up a note about minimizing the number of exceptions reaching the handler - hope that's OK.

jlprat commented on docs/src/main/paradox/java/http/client-side/client-https-support.md:
```diff
@@ -93,7 +93,7 @@ It is specified as part of [RFC 6066](https://tools.ietf.org/html/rfc6066#page-6
 ### Disabling TLS security features, at your own risk
 
-@@@ warning
+``
```
I guess this one slipped through. The warning is good here :)

jlprat commented on docs/src/main/paradox/java/http/routing-dsl/exception-handling.md:
```diff
@@ -37,3 +37,34 @@ The second case allows you to restrict the applicability of your handler to cert
+reaching the exception handler.
+
+To understand the performance implications of (mis-)using exceptions,
+have a read at this excellent post by A. Shipilёv: [https://shipilev.net/blog/2014/exceptional-performance](The Exceptional Performance of Lil' Exception).
```
Regarding external links, currently all external links in the docs are to specs pages.
@jrudolph Is this kind of extra information something to add on the new "right column" that will appear after the docs redesign?
I remember you mentioning this "column" on the docs that could point to akka docs, or to external links where things are explained in further detail

jlprat replied:
Actually, there are a couple of links that are pointing to personal blogs that explain some topic in depth :smiley:

tkroman replied:
Nice! So if external links are good, is there anything else? Should I squash before the PR is accepted?

jlprat replied:
Two people need to give their thumbs up in order to merge a PR. You have one more to go :wink: Feel free to squash if you want, but you can also wait if any other reviewer has any comment.
You can read the process here: https://github.com/akka/akka-http/blob/master/CONTRIBUTING.md#general-workflow

jonas replied:
The link text and URL needs to be flipped:

```md
[The Exceptional Performance of Lil' Exception](https://shipilev.net/blog/2014/exceptional-performance)
```

You can generate the docs by running `docs/paradox`.

jrudolph replied:
The note is fine for me. I think it is generally good advice not to overuse exceptions.

jlprat commented:
I'm dubious about the link, let's see what others think. Otherwise, LGTM

jonas commented on docs/src/main/paradox/scala/http/routing-dsl/exception-handling.md:
```diff
@@ -37,4 +37,35 @@ Here is an example for wiring up a custom handler via @ref[handleExceptions](dir
+reaching the exception handler.
+
+To understand the performance implications of (mis-)using exceptions,
+have a read at this excellent post by A. Shipilёv: [https://shipilev.net/blog/2014/exceptional-performance](The Exceptional Performance of Lil' Exception).
```
^ `[url](text)` => `[text](url)`

tkroman replied:
Done. (oops :/)

tkroman commented:
Squashed.

jrudolph commented:
:+1: thanks!
